File: RxTango/python/examples/throttle.py
"""Rate-limit a fast polling stream using ops.sample.

Polls double_scalar at a high frequency but only passes the most recent
value downstream every N seconds — decoupling producer rate from consumer rate.

Mirrors `TangoTestThrottle.java` (`throttleLast` → `sample` in RxPY).

Usage:
    python throttle.py [device] [poll-ms] [sample-ms]

    device     defaults to tango://localhost:10000/sys/tg_test/1
    poll-ms    poll interval; defaults to 100 (10 Hz)
    sample-ms  output interval; defaults to 1000 (1 Hz)
"""

import asyncio
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


async def main() -> None:
    device    = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    poll_ms   = int(sys.argv[2]) if len(sys.argv) > 2 else 100
    sample_ms = int(sys.argv[3]) if len(sys.argv) > 3 else 1000

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(
        f"Throttle demo — poll={poll_ms} ms, emit latest every {sample_ms} ms "
        f"(Ctrl+C to stop)\n"
    )
    print(f"  {'sampled value':>20}")
    print("  " + "-" * 22)

    rx.interval(timedelta(milliseconds=poll_ms), scheduler=scheduler).pipe(
        ops.flat_map(lambda _: read_attribute(device, "double_scalar")),
        ops.sample(timedelta(milliseconds=sample_ms), scheduler=scheduler),
    ).subscribe(
        on_next=lambda v: print(f"  {v:>+20.6f}"),
        on_error=lambda e: print(f"  ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/throttle.cpp
/**
 * Rate control — poll fast, display slow.
 *
 * Reads at high rate but throttles display to a lower rate using sample().
 * Mirrors Python's throttle.py and Java's TangoTestThrottle.java.
 *
 * Usage:
 *   ./throttle [device] [attribute] [poll-ms] [display-ms]
 *   defaults: sys/tg_test/1  double_scalar  100  1000
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string device     = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr       = argc > 2 ? argv[2] : "double_scalar";
    const int         poll_ms    = argc > 3 ? std::stoi(argv[3]) : 100;
    const int         display_ms = argc > 4 ? std::stoi(argv[4]) : 1000;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Throttle: polling every " << poll_ms << " ms, displaying every "
              << display_ms << " ms  (Ctrl+C to stop)\n\n";

    // Fast poll → throttle via sample() to display rate
    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(poll_ms))
        .flat_map([device, attr](long) {
            return rxtango::read_attribute<double>(device, attr);
        })
        .sample_with_time(std::chrono::milliseconds(display_ms))
        .subscribe(
            [](double v) { std::cout << "  displayed: " << v << "\n"; },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
